bot/exts/evergreen/branding.py

```diff
@@ -0,0 +1,552 @@
+    midnight = datetime.combine(tomorrow, time(second=5))
+
+    return midnight - now
+
+
+class BrandingManager(commands.Cog):
+    """
+    Manages the guild's branding.
+
+    The purpose of this cog is to help automate the synchronization of the branding
+    repository with the guild. It is capable of discovering assets in the repository
+    via Github's API, resolving download urls for them, and delegating
+    to the `bot` instance to upload them to the guild.
+
+    BrandingManager is designed to be entirely autonomous. Its `daemon` background task awakens
+    once a day (see `time_until_midnight`) to detect new seasons, or to cycle icons within a single
+    season. The daemon can be turned on and off via the `daemon` cmd group. The value set via
+    its `start` and `stop` commands is persisted across sessions. If turned on, the daemon will
+    automatically start on the next bot start-up. Otherwise, it will wait to be started manually.
+
+    All supported operations, e.g. setting seasons, applying the branding, or cycling icons, can
+    also be invoked manually, via the following API:
+
+        branding list
+            - Show all available seasons
+
+        branding set <season_name>
+            - Set the cog's internal state to represent `season_name`, if it exists.
+            - If no `season_name` is given, set chronologically current season.
+            - This will not automatically apply the season's branding to the guild,
+              the cog's state can be detached from the guild.
+            - Seasons can therefore be 'previewed' using this command.
+
+        branding info
+            - View detailed information about resolved assets for current season.
+
+        branding refresh
+            - Refresh internal state, i.e. synchronize with branding repository.
+
+        branding apply
+            - Apply the current internal state to the guild, i.e. upload the assets.
+
+        branding cycle
+            - If there are multiple available icons for current season, randomly pick
+              and apply the next one.
+
+    The daemon calls these methods autonomously as appropriate. The use of this cog
+    is locked to moderation roles. As it performs media asset uploads, it is prone to
+    rate-limits - the `apply` command should be used with caution. The `set` command can,
+    however, be used freely to 'preview' seasonal branding and check whether paths have been
+    resolved as appropriate.
+
+    While the bot is in debug mode, it will 'mock' asset uploads by logging the passed
+    download urls and pretending that the upload was successful. Make use of this
+    to test this cog's behaviour.
+    """
+
+    current_season: t.Type[SeasonBase]
+
+    banner: t.Optional[GithubFile]
+    avatar: t.Optional[GithubFile]
+
+    available_icons: t.List[GithubFile]
+    remaining_icons: t.List[GithubFile]
+
+    should_cycle: t.Iterator
+
+    config_file: Path
+
+    daemon: t.Optional[asyncio.Task]
```

Comment:
I do personally think it is cleaner. In some cases, having both the annotation and definition on one line becomes lengthy and visually cluttered. I find that I'm usually looking for one, or the other. My workflow for larger classes is to first annotate the attrs, and then start implementing. Since I've already decided on the types, my IDE will alert me if I violate a promise I made earlier.

Reply:
Thanks for the explanation.